test/jdk/javax/swing/JButton/bug4323121.java

```diff
@@ -30,50 +30,80 @@
  * @run main bug4323121
  */
 
-import java.awt.Graphics;
 import java.awt.Point;
 import java.awt.Robot;
+import java.awt.event.MouseAdapter;
 import java.awt.event.MouseEvent;
-import java.awt.event.MouseListener;
-import java.awt.event.MouseMotionListener;
+import java.awt.event.WindowAdapter;
+import java.awt.event.WindowEvent;
+import java.util.concurrent.CountDownLatch;
+
 import javax.swing.JButton;
 import javax.swing.JFrame;
 import javax.swing.SwingUtilities;
 
-public class bug4323121 {
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public final class bug4323121 extends MouseAdapter {
 
     static JFrame frame;
-    static testButton button;
-    static volatile Point pt;
-    static volatile int buttonW;
-    static volatile int buttonH;
-    static volatile boolean failed = false;
+    static JButton button;
+
+    static volatile Point buttonCenter;
+
+    private static final CountDownLatch windowGainedFocus = new CountDownLatch(1);
+
+    private static final CountDownLatch mouseEntered = new CountDownLatch(1);
+
+    // Thread-safe by using the mouseEntered latch
+    private static boolean modelArmed;
 
     public static void main(String[] args) throws Exception {
         Robot robot = new Robot();
         robot.setAutoDelay(100);
+
+        final bug4323121 eventHandler = new bug4323121();
         try {
             SwingUtilities.invokeAndWait(() -> {
+                button = new TestButton("gotcha");
+                button.addMouseMotionListener(eventHandler);
+                button.addMouseListener(eventHandler);
+
                 frame = new JFrame("bug4323121");
-                button = new testButton("gotcha");
                 frame.getContentPane().add(button);
+
+                frame.addWindowFocusListener(new WindowAdapter() {
+                    @Override
+                    public void windowGainedFocus(WindowEvent e) {
+                        windowGainedFocus.countDown();
+                    }
+                });
+
                 frame.pack();
                 frame.setLocationRelativeTo(null);
                 frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
                 frame.setVisible(true);
             });
+
+            if (!windowGainedFocus.await(1, SECONDS)) {
```

Comment:
Should we add a small delay before checking?

Reply:
It *has* a delay of 1 second: if the event is not delivered within 1 second, the test will fail; it the event is delivered, the test continues.